JavaScript: Add queries and libraries for supporting flow summaries

xiemaisi opened the pull request:
This PR introduces queries and libraries that can be used to generate and consume flow summaries for npm packages using the command-line tools and QL4E.

It's still quite experimental and unfinished, but I'm opening this PR anyway to serve as a basis for discussion.

See `flow-summaries.rst` included in this PR for a brief overview of the various bits of machinery and how to use them.

asger-semmle commented on javascript/ql/src/Security/Summaries/Shared.qll:
```diff
@@ -0,0 +1,95 @@
+  or
+  spec = "" and
+  not config.isBarrier(_) and
+  not config.isBarrier(_, _)
```
There seems to be negative recursion going through here.

The characteristic predicate of the dataflow configuration in `ImportFromCsv` depends on `configSpec` which depends on the dispatch predicate of `isBarrier`, which depends on the set of dataflow configurations.

xiemaisi replied:
Oh dear, I wonder when that crept in. I'll take a look on Monday.

xiemaisi replied:
Fixed by removing the two negative conjuncts. They weren't a good idea in the first place.

asger-semmle commented on javascript/ql/src/Security/Summaries/ImportFromCsv.qll:
```diff
@@ -0,0 +1,192 @@
+  }
+}
+
+private class AdditionalFlowStepFromSpec extends DataFlow::Configuration {
```
It would be nice with a dedicated `AdditionalFlowStep` class for this instead, like with the sources and sinks.

Sort of related: The flow summaries only seem to track `TaintTracking::Configuration`. Is that deliberate? Is there a reason we can't use `AdditionalTaintStep`?

xiemaisi replied:
Hm, good point. There isn't a fundamental reason to only track `TaintTracking::Configuration`s, but it's probably the case we are most interested in for now, so I'll just switch this to `AdditionalTaintStep`.

xiemaisi replied:
Ugh, except that `AdditionalTaintStep` doesn't currently allow specifying kinds, and it is also `cached` and should only be extended in the standard library. Bit messy, that.

asger-semmle replied:
Ah, yeah the `cached` thing is a deal-breaker. Still, it seems a bit sneaky to have this configuration-class here, so if we could add an uncached class for this that would be great.

asger-semmle commented on javascript/ql/src/Security/Summaries/ExtractFlowStepSummaries.ql:
```diff
@@ -0,0 +1,32 @@
+      p1 = source.getNode().(PortalExitSource).getPortal() and
+      p2 = sink.getNode().(PortalEntrySink).getPortal() and
+      lbl1 = sink.getPathSummary().getStartLabel() and
+      lbl2 = sink.getPathSummary().getStartLabel() and
```
Copy-pasta? The two labels will always be the same, won't they?

xiemaisi replied:
Indeed; my least favourite kind of pasta.

xiemaisi commented:
>  It seems we could safely add ImportFromCsv in javascript.qll

Yes, I think that would be a good idea, but I'll need to check that it doesn't unduly affect performance when there aren't any summaries to import.

xiemaisi commented:
Rebased to resolve auto-formatting conflicts.

I've experimented with ways of reducing the performance penalty from including `ImportFromCSV` in the standard library, but couldn't bring the overhead down to acceptable levels, so I've reverted that commit. With the revert, no performance or result changes were observed on a full dist-compare run.

This means that flow-summary support is opt-in and requires a local change to `javascript.qll`, but perhaps that's good enough for now.

xiemaisi commented:
@asger-semmle: I had to add a commit to resolve a semantic merge conflict (`document.cookie`, which was used in one of the tests) is no longer considered a source of remote flow. Could you take a look?
